File: app/src/main/java/com/dani_chuks/andeladeveloper/starwars/home/HomeViewModelFactory.java
package com.dani_chuks.andeladeveloper.starwars.home;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.dani_chuks.andeladeveloper.starwars.dagger.ISchedulerProvider;
import com.dani_chuks.andeladeveloper.starwars.home.HomeViewModelInteractor;

public class HomeViewModelFactory implements ViewModelProvider.Factory {

    @NonNull
    private final ISchedulerProvider schedulerProvider;
    @NonNull
    private final HomeViewModelInteractor viewModelInteractor;

    public HomeViewModelFactory(@NonNull final HomeViewModelInteractor viewModelInteractor,
                                @NonNull final ISchedulerProvider schedulerProvider) {
        this.schedulerProvider = schedulerProvider;
        this.viewModelInteractor = viewModelInteractor;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull final Class<T> modelClass) {
        if (modelClass.isAssignableFrom(HomeViewModel.class)) {
            return (T) new HomeViewModel(viewModelInteractor, schedulerProvider);
        }
        throw new IllegalArgumentException("Unknown ViewModel Class");
    }
}

File: app/src/main/java/com/dani_chuks/andeladeveloper/starwars/home/HomeViewModelFactory.kt
package com.dani_chuks.andeladeveloper.starwars.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.dani_chuks.andeladeveloper.starwars.di.IDispatcherProvider
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import javax.inject.Inject

@ExperimentalCoroutinesApi
@FlowPreview
@Suppress("UNCHECKED_CAST")
class HomeViewModelFactory @Inject constructor(private val homeModelStore: HomeModelStore)
    : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(HomeViewModel::class.java))
            return HomeViewModel(homeModelStore) as T
        throw IllegalArgumentException("Unknown ViewModel Class")
    }
}
